### PR TITLE
[5.7] [WIP] bug fix for app()->call()

### PR DESCRIPTION
**Pull request by imanghafoori1**
Some changes in the BoundMethod class to make a fix addressing these issues:

https://github.com/laravel/framework/issues/26851
https://github.com/laravel/framework/issues/22687
https://github.com/laravel/framework/issues/26837

Following the my previous bug fix https://github.com/laravel/framework/pull/26852 which caused a new nasty bug and eventually got reverted...
I created an other PR to add the missing tests to cover more cases https://github.com/laravel/framework/pull/26897 

After one sleepless night I am coming up with an other fundamental solution for making it really fixed.

I know I am asked not to make a PR again for that feature but please let me try again.
This time much more vigilant, with more tests.

I will add more tests tomorrow to be really sure that no more revert will happen

### CODE REVIEW

**Comment by deleugpn** [comment]
You have a better chance by targeting the master branch and fixing the issue for the next release. It gives more time for thorough test.

**Comment by taylorotwell** [comment]
Absolutely no way this gets merged on 5.7.

**Comment by imanghafoori1** [comment]
@taylorotwell
Ok I just continue working on it.
You may consider merging it for 5.8 or ignore it anyway.
